prac_07/project_management.py

```diff
@@ -0,0 +1,177 @@
+    return new_completion_percentage
+
+def load_projects():
+    try:
+        with open("projects.dat", "rb") as file:
+            projects = pickle.load(file)
+    except FileNotFoundError:
+        projects = []
+    return projects
+
+
+def save_projects(projects):
+    with open("projects.dat", "wb") as file:
+        pickle.dump(projects, file)
+
+
+def display_projects(projects):
+    unfinished = sorted([p for p in projects if p.completion < 100])
+    finished = sorted([p for p in projects if p.completion == 100])
+
+    print("Unfinished Projects:")
+    if unfinished:
+        for p in unfinished:
+            print(p)
+    else:
+        print("No unfinished projects.")
+    print()
+
+    print("Finished Projects:")
+    if finished:
+        for p in finished:
+            print(p)
+    else:
+        print("No finished projects.")
+    print()
+
+def filter_projects_by_date(projects):
+    date_str = input("Enter start date (mm/dd/yyyy): ")
+    try:
+        date = datetime.strptime(date_str, "%m/%d/%Y")
+    except ValueError:
+        print("Invalid date format. Please use mm/dd/yyyy.")
+        return
+
+    filtered = sorted([p for p in projects if p.start_date >= date], key=lambda p: p.start_date)
+
+    print("Filtered Projects:")
+    if filtered:
+        for p in filtered:
+            print(p)
+    else:
+        print("No projects found for that date or later.")
+    print()
+
+def add_project(projects):
+    name = input("Enter project name: ")
+    priority = int(input("Enter project priority (1-5): "))
+    date_str = input("Enter start date (mm/dd/yyyy): ")
+    try:
+        date = datetime.strptime(date_str, "%m/%d/%Y")
+    except ValueError:
+        print("Invalid date format. Please use mm/dd/yyyy.")
+        return
+
+    projects.append(projects(name, priority, date))
+    print("Project added.")
+    print()
+
+def update_project(projects):
+    print("Select a project to update:")
+    for i, p in enumerate(projects):
+        print(f"{i + 1}. {p}")
+    try:
+        selection = int(input("Enter project number: "))
+        project = projects[selection - 1]
+    except (ValueError, IndexError):
+        print("Invalid selection.")
+        return
+    completion = input(f"Enter completion percentage ({project.completion}%): ")
+    if completion:
+        try:
+            completion = int(completion)
+            if completion < 0 or completion > 100:
+                raise ValueError
+        except ValueError:
+            print("Invalid completion percentage.")
+            return
+        project.completion = completion
+
+    priority = input(f"Enter priority ({project.priority}): ")
+    if priority:
+        try:
+            priority = int(priority)
+            if priority < 1 or priority > 5:
+                raise ValueError
+        except ValueError:
+            print("Invalid completion percentage.")
+            return
+        project.priority = priority
+
+        main()
+
+
+
```

Comment:
The expression is clear, and the code is clean and flowing.

Reply:
Yeah, and also the code has been well written and organized, which make us easy to understand 👍